synse/http_integration_test.go

```diff
@@ -0,0 +1,551 @@
+		assert.NoError(t, err)
+	}()
+
+	for {
+		var done bool
+		select {
+		case read, open := <-reads:
+			if !open {
+				done = true
+				break
+			}
+
+			assert.NotEmpty(t, read.Device)
+			assert.NotEmpty(t, read.Timestamp)
+			// assert.NotEmpty(t, read.Type) // FIXME - fan and airflow types are empty?
+			assert.NotEmpty(t, read.DeviceType)
+
+			// led and lock devices don't produce unit output
+			if read.DeviceType == "led" || read.DeviceType == "lock" {
+				assert.Empty(t, read.Unit.Name)
+				assert.Empty(t, read.Unit.Symbol)
+			} else {
+				assert.NotEmpty(t, read.Unit.Name)
+				assert.NotEmpty(t, read.Unit.Symbol)
+			}
+
+			// NOTE - read.Value could be 0 so no need to check that
+
+		case <-time.After(2 * time.Second):
+			// if the test does not complete after 2s, error.
+			t.Fatal("timeout: failed getting readcache data from channel")
+		}
+
+		if done {
+			break
+		}
+	}
+}
+
+func TestIntegration_WriteAsync(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping integration test")
+	}
+
+	client, err := NewHTTPClientV3(&Options{
+		Address: "localhost:5000",
+	})
+	assert.NotNil(t, client)
+	assert.NoError(t, err)
+
+	// collect all writable devices.
+	// FIXME - refer to #69, can only query one type atm.
+	opts := scheme.ScanOptions{
+		Tags: []string{
+			"system/type:fan",
+			// "system/type:led",
+			// "system/type:lock",
+			// "system/type:power",
+		},
+	}
+	devices, err := client.Scan(opts)
+	assert.NoError(t, err)
+	assert.Equal(t, 1, len(devices))
+
+	for _, device := range devices {
+		assert.NotEmpty(t, device.ID)
+
+		writeData := []scheme.WriteData{
+			{Action: "speed", Data: "101"}, // FIXME - if Data is int, get 500 from server.
+		}
+		writes, err := client.WriteAsync(device.ID, writeData)
+		assert.NoError(t, err)
+
+		assert.Equal(t, 1, len(writes))
+
+		for _, write := range writes {
+			assert.NotEmpty(t, write.ID)
+			assert.NotEmpty(t, write.Device)
+			assert.Equal(t, "speed", write.Context.Action)
+			// FIXME - reflected data is not decoded yet
+			// assert.Equal(t, "101"), write.Context.Data)
+			assert.Empty(t, write.Context.Transaction)
+			assert.NotEmpty(t, write.Timeout)
+		}
+	}
+}
+
+func TestIntegration_WriteSync(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping integration test")
+	}
+
+	client, err := NewHTTPClientV3(&Options{
+		Address: "localhost:5000",
+	})
+	assert.NotNil(t, client)
+	assert.NoError(t, err)
+
+	// collect all writable devices.
+	// FIXME - refer to #69, can only query one type atm.
+	opts := scheme.ScanOptions{
+		Tags: []string{
+			"system/type:fan",
+			// "system/type:led",
+			// "system/type:lock",
+			// "system/type:power",
+		},
+	}
+	devices, err := client.Scan(opts)
+	assert.NoError(t, err)
+	assert.Equal(t, 1, len(devices))
+
+	for _, device := range devices {
+		assert.NotEmpty(t, device.ID)
+
+		writeData := []scheme.WriteData{
+			{Action: "speed", Data: "101"}, // FIXME - if Data is int, get 500 from server.
+		}
```

Comment:
this works right now since we are only writing to fan devices. if we include different kinds, this will cause failures for those devices which do not support this action. I think instead of trying to write to all devices, we should focus on writing to one device and verifying all the data from that